ipam/paxos/paxos_test.go

```diff
@@ -0,0 +1,351 @@
+	// this node, then a different consensus can occur later on.
+	// If a tree falls with no one to hear it, does it make a
+	// sound?
+	node.firstConsensus = AcceptedValue{}
+}
+
+func (m *Model) pickNode() *TestNode {
+	for {
+		node := &m.nodes[m.r.Intn(len(m.nodes))]
+		if !node.isolated {
+			return node
+		}
+	}
+}
+
+func (m *Model) simulate(params *TestParams) bool {
+	nodesLeft := uint(len(m.nodes))
+	restarts := uint(0)
+
+	for step := 0; step < 1000000; step++ {
+		if len(m.readyLinks) == 0 {
+			// Everything has converged.  This can be
+			// because consensus was reached, or because
+			// consensus became impossible, e.g. everyone
+			// promised on a particular proposal, but then
+			// the proposer restarted or was isolated.  So
+			// we detect the latter cases and force a new
+			// proposal
+			for i := range m.nodes {
+				ok, _ := m.nodes[i].Consensus()
+				if ok {
+					return true
+				}
+			}
+
+			node := m.pickNode()
+			node.Propose()
+			m.nodeChanged(node)
+		}
+
+		// Pick a ready link at random
+		i := m.r.Intn(len(m.readyLinks))
+		link := m.readyLinks[i]
+		if link.ready != i {
+			m.t.Fatal("Link in readyLinks was not ready")
+		}
+
+		// gossip across link
+		node := link.to
+		if node.Update(link.from.GossipState()) {
+			node.Think()
+
+			if !node.firstConsensus.Origin.valid() {
+				ok, val := node.Consensus()
+				if ok {
+					node.firstConsensus = val
+				}
+			}
+
+			m.nodeChanged(node)
+		}
+
+		m.unreadyLink(link)
+
+		// Re-propose?
+		if m.r.Float32() < params.reproposeProb {
+			node := m.pickNode()
+			node.Propose()
+			m.nodeChanged(node)
+		}
+
+		// Isolate?
+		if nodesLeft > m.quorum && m.r.Float32() < params.isolateProb {
+			m.isolateNode(m.pickNode())
+			nodesLeft--
+
+			// We isolated a node, so get another node to
+			// re-propose.  In reality the lack of
+			// consensus would be detected via a timeout
+			node := m.pickNode()
+			node.Propose()
+			m.nodeChanged(node)
+		}
+
+		// Restart?
+		if restarts < m.quorum && m.r.Float32() < params.restartProb {
+			restarts++
+			node := m.pickNode()
+			m.restart(node)
+			m.nodeChanged(node)
+		}
+	}
+
+	return false
+}
+
+func (m *Model) dump() {
+	for i := range m.nodes {
+		node := &m.nodes[i]
+		fmt.Println(node.id)
+		for n, claims := range node.knows {
+			fmt.Printf("\t%d %v\n", n, claims)
+		}
+	}
+}
+
+// Validate the final model state
+func (m *Model) validate() {
+	var origin ProposalID
+
+	for i := range m.nodes {
+		ok, val := m.nodes[i].Consensus()
+		if !ok {
+			//m.dump()
+			m.t.Fatal("Node doesn't know about consensus")
+		}
+
+		if m.nodes[i].firstConsensus.Origin != val.Origin {
+			//m.dump()
+			m.t.Fatal("Consensus mismatch")
+		}
+
+		if i == 0 {
+			origin = val.Origin
+		} else if val.Origin != origin {
+			m.t.Fatal("Node disagrees about consensus")
+		}
+	}
+}
+
+func TestPaxos(t *testing.T) {
+	r := rand.New(rand.NewSource(time.Now().UnixNano()))
+
+	params := TestParams{
+		nodeCount:     10,
```

Comment:
Why 10? To allow for three partitions, with non-trivial (i.e. not 0,1,2) node counts in each?